pyvo/dal/adhoc.py

```diff
@@ -167,48 +170,127 @@ def get_adhocservice_by_id(self, id_):
-                current_batch = self.query.execute(post=True)
-                current_ids = list(OrderedDict.fromkeys(
-                    [_ for _ in current_batch.to_table()['ID']]))
-                if not current_ids:
-                    raise DALServiceError(
-                        'Could not retrieve datalinks for: {}'.format(
-                            ', '.join([_ for _ in remaining_ids])))
-                batch_size = len(current_ids)
-            id1 = current_ids.pop(0)
-            processed_ids.append(id1)
-            remaining_ids.remove(id1)
-            yield current_batch.clone_byid(
-                id1,
-                original_row=row)
+        Parameters
+        ----------
+
+        preserve_order : bool
+            True to return the datalinks keeping the order of the current rows.
+            NOTE: There might be a performance penalty for keeping the order as
+            one request per row is sent to the service. When the order of the
+            datalinks is not important, the execution is optimized to query the
+            service in batches.
+
+        """
+        def _get_results_tb(rows, dl_batch_tb):
+            # Creates a new DL result table with the given rows as the results data
+            # and the dl_batch_tb as the template for both table fields and other resources
+            tb = VOTableFile()
+            new_table = TableElement(tb)
+            new_table.fields.extend(dl_batch_tb.get_first_table().fields)
+            new_table.create_arrays(len(rows))
+            for index, row in enumerate(rows):
+                new_table.array[index] = row
+            results_resource = Resource()
+            results_resource.type = "results"
+            results_resource.tables.append(new_table)
+            tb.resources.append(results_resource)
+            # now add all the other resources from the dl_batch_tb
+            for resource in dl_batch_tb.resources:
+                if resource.type != "results":
+                    tb.resources.append(resource)
+            return tb
+
+        if preserve_order:
+            # one request at the time to preserve order
+            for row in self:
+                self.query = DatalinkQuery.from_resource(
+                    row, self._datalink, session=self._session, original_row=None)
+                yield DatalinkResults(
+                    self.query.execute(post=True).votable,
+                    original_row=row)
+            return
+
+        # map of IDs to original rows needed to map the results back to the
+        # original rows
+        original_rows = {}
+        input_params = _get_input_params_from_resource(self._datalink)
+        for name, input_param in input_params.items():
+            for row in self:
+                try:
+                    if row[input_param.ref]:
+                        original_rows[row[input_param.ref]] = row
+                except KeyError:
```

Comment:
I suspect our futures selves will be grateful if we explain in a brief comment how the original_row thing works in total (I have to admit I'm not quite sure myself) and in particular why a KeyError (from row[input_param.ref], right?) is ignorable here.

Reply:
The block creates a map of DL link ID->original row by using the `ref` field of the input parameter of the service descriptor. I hope that the description is a bit clearer now but I stand to be corrected.
Not finding the `ref` in the original results table should never happen so following the example of `DatalinkQuery.from_resource` I remove that check.